src/main/haskell/matching-logic/src/Kore/MatchingLogic/HilbertProof.hs

```diff
@@ -0,0 +1,65 @@
+module Kore.MatchingLogic.HilbertProof
+  (Proof(index,claims,derivations)
+  ,ProofSystem(..)
+  ,emptyProof
+  ,add
+  ,derive
+  ,renderProof
+  )   where
+import Data.List(lookup,delete,nub)
+import Control.Monad(guard)
+import Data.Map.Strict(Map)
+import qualified Data.Map.Strict as Map
+import Data.Sequence(Seq,(|>))
+import qualified Data.Sequence as Seq
+import Data.Foldable
+
+data Proof ix rule formula =
+  Proof
+  { index :: Map ix (Int,formula)
+  , claims :: Seq (ix,formula)
+  , derivations :: Map ix (rule,[(ix,formula)])
+  }
+  deriving (Show)
+
+emptyProof :: Proof ix rule formula
+emptyProof = Proof Map.empty Seq.empty Map.empty
+
+add :: (Ord ix)
+    => Proof ix rule formula -> ix -> formula -> Maybe (Proof ix rule formula)
+add proof ix formula
+  | not (Map.member ix (index proof)) = Just $
+    proof { index = Map.insert ix (Seq.length (claims proof), formula) (index proof)
+          , claims = claims proof |> (ix,formula)
+          , derivations = derivations proof
+          }
+  | otherwise = Nothing
+
+renderProof :: (Ord ix, Show ix, Show rule, Show formula)
+            => Proof ix rule formula -> String
+renderProof proof = unlines
+  [show ix++" : "++show formula++
+   case Map.lookup ix (derivations proof) of
+     Nothing -> ""
+     Just (rule,[]) -> " by "++show rule
+     Just (rule,arguments) -> " by "++show rule++" from "++
+       unwords (map (show . fst) arguments)
+  | (ix,formula) <- toList (claims proof)]
+
+class Eq formula => ProofSystem rule formula | rule -> formula where
+  checkDerivation :: rule -> formula -> [formula] -> Bool
```

Comment:
Just checking that my understanding is correct: since checkDerivation uses a formula list, and not a list of indexes, this allows applying inference rules on unproven formulas, right?

Reply:
Yes, that's the design. Unproved formulas include the axioms we want to assume, and also the API is designed to allow adding a derivation for a previously unproved formula.